Tribler/Core/Libtorrent/LibtorrentMgr.py

```diff
@@ -102,7 +105,21 @@ def initialize(self):
         self.register_task(u'task_cleanup_metacache',
                            LoopingCall(self._task_cleanup_metainfo_cache)).start(60, now=True)
 
-    def shutdown(self):
+    def shutdown(self, timeout=30):
+        self.tribler_session.notify_shutdown_state("Shutting down Libtorrent Manager...")
+        # If libtorrent session has pending disk io, wait until timeout (default: 30 seconds) to let it finish.
+        # In between ask for session stats to check if state is clean for shutdown.
+        if not self.is_shutdown_ready() and timeout > 5:
+            self.tribler_session.notify_shutdown_state("Waiting for Libtorrent to finish...")
+            self.post_session_stats()
+
+            def schedule_shutdown():
+                self.register_anonymous_task("schedule_shutdown",
+                                             reactor.callLater(5, lambda t=timeout-5: self.shutdown(timeout=t)))
+
+            reactor.callFromThread(schedule_shutdown)
```

Comment:
Please see https://github.com/Tribler/py-ipv8/pull/363/files for the certified IPv8 way of doing this.